src/Repository.js

```diff
@@ -118,8 +120,22 @@ export default class Repository {
   }
 
   buildPackageGraph() {
```

Comment:
We should pass an options object derived from `this.options` when it is called from [`Command.js`](https://github.com/lerna/lerna/blob/5ea7e2d9a216bb2274e217401eb47db8661f6068/src/Command.js#L306) and use those properties instead of `this.lernaJson`. This allows us to support CLI flags (`--use-git-version`, `--git-version-prefix`) and eventually argument validation with yargs.

```diff
diff --git a/src/Command.js b/src/Command.js
index e5bb0c4..032cd01 100644
--- a/src/Command.js
+++ b/src/Command.js
@@ -288,7 +288,7 @@ export default class Command {
   }
 
   runPreparations() {
-    const { scope, ignore, registry, since } = this.options;
+    const { scope, ignore, registry, since, useGitVersion, gitVersionPrefix } = this.options;
 
     if (scope) {
       log.info("scope", scope);
@@ -303,7 +303,7 @@ export default class Command {
     }
 
     try {
-      this.repository.buildPackageGraph();
+      this.repository.buildPackageGraph({ useGitVersion, gitVersionPrefix });
       this.packages = this.repository.packages;
       this.packageGraph = this.repository.packageGraph;
       this.filteredPackages = PackageUtilities.filterPackages(this.packages, { scope, ignore });
```
(Sorry for the scattered review)

Reply:
Any suggestion what would be the best way to deal with `buildPackageGraph` being called from
[`packages` and `packageGraph` getters](https://github.com/lerna/lerna/blob/master/src/Repository.js#L70-L81)?

For now I have removed those implicit calls, since otherwise I think we would need to set current command options on the repository and then pass them:
```
  get packageGraph() {
    if (!this._packageGraph) {
      this.buildPackageGraph(this.options); // this would have to be set eg. from command which looks a bit messy...
    }

    return this._packageGraph;
  }
```

Unless there's an acces to currently running command from Repository?

Reply:
Yeah, I totally understand the mixed feelings. I have them myself, about the Repository class in general. It's doing too many disparate things for my taste, but it's been low on the priority list to refactor...

So: It occurs to me that Repository itself _shouldn't_ be storing `packages` and `packageGraph` in getters since there are literally two places in the entire codebase that they are referenced:
1. in Command, right after the `buildPackageGraph()` is called
2. in [UpdatedPackagesCollector](https://github.com/lerna/lerna/blob/b350871597d46da602eb54733bd398ff0c9b8d07/src/UpdatedPackagesCollector.js#L40), where it violates separation of concern by reaching into (sub-)properties of the command instance just to retrieve mostly-static (i.e., unchanging) values.

Here's what we can do: Move the creation of both the `packages` list and the `packageGraph` into Command, where it belongs, and completely remove the `get packages` and `get packageGraph` from Repository.

This means all the additional logic added here will be contained inside the place where `this.options` already exists. Something like this:
```diff
diff --git a/src/Command.js b/src/Command.js
index 58659bb..aa4bc7f 100644
--- a/src/Command.js
+++ b/src/Command.js
@@ -294,7 +294,8 @@ export default class Command {
   }
 
   runPreparations() {
-    const { scope, ignore, registry, since } = this.options;
+    const { rootPath, packageConfigs } = this.repository;
+    const { scope, ignore, registry, since, useGitVersion, gitVersionPrefix } = this.options;
 
     if (scope) {
       log.info("scope", scope);
@@ -309,10 +310,22 @@ export default class Command {
     }
 
     try {
-      this.repository.buildPackageGraph();
-      this.packages = this.repository.packages;
-      this.packageGraph = this.repository.packageGraph;
-      this.filteredPackages = PackageUtilities.filterPackages(this.packages, { scope, ignore });
+      const versionParser = useGitVersion && new GitVersionParser(gitVersionPrefix);
+      const packages = PackageUtilities.getPackages({ rootPath, packageConfigs });
+      const packageGraph = PackageUtilities.getPackageGraph(packages, false, versionParser);
+
+      if (useGitVersion) {
+        packages.forEach((pkg) => {
+          pkg.versionSerializer = new VersionSerializer({
+            graphDependencies: packageGraph.get(pkg.name).dependencies,
+            versionParser,
+          });
+        });
+      }
+
+      this.packages = packages;
+      this.packageGraph = packageGraph;
+      this.filteredPackages = PackageUtilities.filterPackages(packages, { scope, ignore });
 
       // The UpdatedPackagesCollector requires that filteredPackages be present prior to checking for
       // updates. That's okay because it further filters based on what's already been filtered.
diff --git a/src/Repository.js b/src/Repository.js
index 8316083..7f1e138 100644
--- a/src/Repository.js
+++ b/src/Repository.js
@@ -8,7 +8,6 @@ import semver from "semver";
 
 import dependencyIsSatisfied from "./utils/dependencyIsSatisfied";
 import Package from "./Package";
-import PackageUtilities from "./PackageUtilities";
 
 const DEFAULT_PACKAGE_GLOB = "packages/*";
 
@@ -67,22 +66,6 @@ export default class Repository {
       .map(parentDir => path.resolve(this.rootPath, parentDir));
   }
 
-  get packages() {
-    if (!this._packages) {
-      this.buildPackageGraph();
-    }
-
-    return this._packages;
-  }
-
-  get packageGraph() {
-    if (!this._packageGraph) {
-      this.buildPackageGraph();
-    }
-
-    return this._packageGraph;
-  }
-
   get packageJson() {
     if (!this._packageJson) {
       try {
@@ -117,11 +100,6 @@ export default class Repository {
     return this.version === "independent";
   }
 
-  buildPackageGraph() {
-    this._packages = PackageUtilities.getPackages(this);
-    this._packageGraph = PackageUtilities.getPackageGraph(this._packages);
-  }
-
   hasDependencyInstalled(depName, version) {
     log.silly("hasDependencyInstalled", "ROOT", depName, version);
 
diff --git a/src/UpdatedPackagesCollector.js b/src/UpdatedPackagesCollector.js
index 67d8ba5..5e51b60 100644
--- a/src/UpdatedPackagesCollector.js
+++ b/src/UpdatedPackagesCollector.js
@@ -31,7 +31,7 @@ export default class UpdatedPackagesCollector {
     this.logger = command.logger;
     this.repository = command.repository;
     this.packages = command.filteredPackages;
-    this.packageGraph = command.repository.packageGraph;
+    this.packageGraph = command.packageGraph;
     this.options = command.options;
   }
 
diff --git a/test/Repository.js b/test/Repository.js
index 1c21bfc..8b01475 100644
--- a/test/Repository.js
+++ b/test/Repository.js
@@ -142,32 +142,6 @@ describe("Repository", () => {
     });
   });
 
-  describe("get .packages", () => {
-    it("returns the list of packages", () => {
-      const repo = new Repository(testDir);
-      expect(repo.packages).toEqual([]);
-    });
-
-    it("caches the initial value", () => {
-      const repo = new Repository(testDir);
-      expect(repo.packages).toBe(repo.packages);
-    });
-  });
-
-  describe("get .packageGraph", () => {
-    it("returns the graph of packages", () => {
-      const repo = new Repository(testDir);
-      expect(repo.packageGraph).toBeDefined();
-      expect(repo.packageGraph).toHaveProperty("nodes", []);
-      expect(repo.packageGraph).toHaveProperty("nodesByName", {});
-    });
-
-    it("caches the initial value", () => {
-      const repo = new Repository(testDir);
-      expect(repo.packageGraph).toBe(repo.packageGraph);
-    });
-  });
-
   describe("get .packageJson", () => {
     afterEach(() => {
       readPkg.sync = readPkgSync;
 ```